ci/tasks/configure_libvirt_gpu.py

```diff
@@ -0,0 +1,103 @@
+import argparse
+import logging
+import os
+import subprocess
+import time
+import xml.etree.ElementTree as ET
+
+logger = logging.getLogger('Libvirt-Attach')
+
+
+class ConfigureLibvirtDevice:
+
+    def shutdown_domain(self, domain):
+        if 'shut off' not in self._get_domain_status(domain):
+            logger.info(f'Shutting down {domain}')
+            self._run_cmd(f'virsh shutdown {domain}')
+
+            if not self._wait_for_status(domain, 'shut off'):
+                logger.warning(f'{domain} still not off, forcing off')
+                self._run_cmd(f'virsh destroy {domain}')
+
+    def detach_device(self, device_id):
+        logger.info(f'Detaching {device_id} from host')
+        self._run_cmd(f'virsh nodedev-dettach {device_id}')
+
+    def attach_device(self, domain, device_id):
+        logger.info(f'Attaching {device_id} to {domain}')
+        device_addresses = configure._get_device_addresses(device_id)
+        config_file = configure._write_config_file(f'gpu_{device_id}', device_addresses)
+
+        self._run_cmd(f'virsh attach-device {domain} --file {config_file} --config')
+
+    def start_domain(self, domain):
+        logger.info(f'Starting {domain}')
+        self._run_cmd(f'virsh start {domain}')
+
+    def _get_device_addresses(self, device_id):
+        output = self._run_cmd(f'virsh nodedev-dumpxml {device_id}')
+        root = ET.fromstring(output)
+
+        return root.findall("./capability/iommuGroup/address")
+
+    def _get_domain_status(self, domain):
+        return self._run_cmd(f'virsh domstate {domain}')
+
+    def _run_cmd(self, cmd):
+        logger.debug(cmd)
+
+        proc = subprocess.run(cmd,
+                            encoding='utf8',
+                            shell=True,
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.STDOUT)
+        if proc.returncode != 0:
+            raise Exception(f'Received exit code {proc.returncode} while running command {cmd}\n{proc.stdout}')
+
+        return proc.stdout
+
+    def _wait_for_status(self, domain, status, timeout=60):
+        current_status = self._get_domain_status(domain)
+
+        while status not in current_status and timeout > 0:
+            timeout -= 1
+            time.sleep(1)
+            current_status = self._get_domain_status(domain)
+
+        return status in current_status
+
+    def _write_config_file(self, name, device_addresses):
+        config_path = os.path.join(os.getcwd(), f'{name}.xml')
```

Comment:
I'd generally tend to prefer to put this file in a tempfile using `tempfile.NamedTemporaryFile` instead of assuming that the cwd is writable.  That would prevent stomping on an existing file as well.